src/command/shim.rs

```diff
@@ -0,0 +1,225 @@
+use std::ffi::OsStr;
+use std::fmt::{self, Display, Formatter};
+use std::fs;
+use std::path::PathBuf;
+
+use notion_core::session::{ActivityKind, Session};
+use notion_core::{path, shim, style};
+use notion_fail::{Fallible, ResultExt};
+use semver::VersionReq;
+
+use Notion;
+use command::{Command, CommandName, Help};
+
+#[derive(Debug, Deserialize)]
+pub(crate) struct Args {
+    arg_shimname: Option<String>,
+    flag_delete: bool,
+    flag_verbose: bool,
+}
+
+pub(crate) enum Shim {
+    Help,
+    List(bool),
+    Create(String, bool),
+    Delete(String, bool),
+}
+
+enum ShimKind {
+    Local(PathBuf),
+    Global(PathBuf),
+    System,
+    WillInstall(VersionReq),
+    Unimplemented,
+}
+
+impl Display for ShimKind {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
+        let s = match self {
+            &ShimKind::Local(ref path) => format!("{}", path.to_string_lossy()),
+            &ShimKind::Global(ref path) => format!("{}", path.to_string_lossy()),
+            &ShimKind::System => format!("[system]"),
+            &ShimKind::WillInstall(ref version) => format!("[will install version {}]", version),
+            &ShimKind::Unimplemented => format!("[shim not implemented!]"),
+        };
+        f.write_str(&s)
+    }
+}
+
+impl Command for Shim {
+    type Args = Args;
+
+    const USAGE: &'static str = "
+Manage Notion shims for 3rd-party executables
+
+Usage:
+    notion shim [options]
+    notion shim <shimname> [options]
+
+Options:
+    -d, --delete   Delete 3rd-party shim
+    -v, --verbose  Verbose output
+    -h, --help     Display this message
+
+";
+
+    fn help() -> Self {
+        Shim::Help
+    }
+
+    fn parse(
+        _: Notion,
+        Args {
+            arg_shimname,
+            flag_delete,
+            flag_verbose,
+        }: Args,
+    ) -> Fallible<Self> {
+        Ok(if let Some(shim_name) = arg_shimname {
+            if flag_delete {
+                Shim::Delete(shim_name, flag_verbose)
+            } else {
+                Shim::Create(shim_name, flag_verbose)
+            }
+        } else {
+            Shim::List(flag_verbose)
+        })
+    }
+
+    fn run(self, session: &mut Session) -> Fallible<bool> {
+        session.add_event_start(ActivityKind::Shim);
+
+        let result = match self {
+            Shim::Help => Help::Command(CommandName::Shim).run(session),
+            Shim::List(verbose) => list(session, verbose),
+            Shim::Create(shim_name, verbose) => create(session, shim_name, verbose),
+            Shim::Delete(shim_name, verbose) => delete(session, shim_name, verbose),
+        };
+        session.add_event_end(ActivityKind::Shim, 0);
+        result
+    }
+}
+
+fn list(session: &mut Session, verbose: bool) -> Fallible<bool> {
```

Comment:
This function is hard to read. Can you break out some of the nested loops into helper functions?

Reply:
Done!